source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -73,76 +238,1816 @@ namespace rex
 
     namespace backend
     {
+      static constexpr s32 s_swapchain_buffer_count = 2;
+
+      static constexpr s32 s_max_color_targets = 8;
+      static constexpr s32 s_max_depth_targets = 1;
+
+      static constexpr s32 s_constant_buffer_min_allocation_size = 256;
+
+      static constexpr s32 s_rtv_descriptor_count = s_max_color_targets;
+      static constexpr s32 s_dsv_descriptor_count = s_max_depth_targets;
+      static constexpr s32 s_cbv_descriptor_count = 128;
+
+      static constexpr s32 s_num_frame_resources = 3;
+
+      struct Frame
+      {
+          Frame(s32 idx, ResourceSlot s)
+              :index(idx)
+              ,slot(s)
+          {}
+
+          s32 index;
+          ResourceSlot slot;
+
+          rsl::vector<ResourceSlot> commited_resources;
+          rsl::unordered_map<const ResourceSlot*, s32> active_constant_buffers;
+      };
+
+      struct FrameContext
+      {
+          FrameContext(s32 maxFrameResources)
+              : frame_resources()
+              , curr_frame_resource(nullptr)
+              , curr_frame_resource_index(REX_INVALID_INDEX)
+          {
+              REX_ASSERT_X(maxFrameResources > 0, "A minimum of one frame has to be created in order to render anything.");
+
+              frame_resources.reserve(maxFrameResources);
+          }
+
+          rsl::vector<Frame>    frame_resources;
+
+          const ResourceSlot*   curr_frame_resource;
+          s32                   curr_frame_resource_index;
+      };
+
       struct DirectXContext
       {
         wrl::com_ptr<ID3D12Device> device = nullptr;
+        wrl::com_ptr<ID3D12Fence> fence   = nullptr;
+
+        u64 current_fence;
+
+        s32 rtv_desc_size         = 0;
+        s32 dsv_desc_size         = 0;
+        s32 cbv_srv_uav_desc_size = 0;
+
+        DXGI_FORMAT back_buffer_format   = DXGI_FORMAT_R8G8B8A8_UNORM;
+        DXGI_FORMAT depth_stencil_format = DXGI_FORMAT_D24_UNORM_S8_UINT;
+
+        bool msaa_state  = false;
+        s32 msaa_quality = 0;
+
+        wrl::com_ptr<ID3D12CommandQueue> command_queue         = nullptr;
+        wrl::com_ptr<ID3D12GraphicsCommandList> command_list   = nullptr;
+        wrl::com_ptr<ID3D12CommandAllocator> command_allocator = nullptr;
+
+        wrl::com_ptr<IDXGISwapChain> swapchain = nullptr;
+
+        rsl::unordered_map<D3D12_DESCRIPTOR_HEAP_TYPE, wrl::com_ptr<ID3D12DescriptorHeap>> descriptor_heap_pool;
+        rsl::unordered_map<rsl::hash_result, wrl::com_ptr<ID3D12PipelineState>> pipeline_state_objects;
```

Comment:
why does this take a hash_result as key? it's an unordered_map, so you don't need to hash things yourself

Reply:
rsl::unordered_set is not implement

Reply:
you're still hashing things yourself, I'll make an issue for it

Reply:
https://github.com/New-Creta/Rex/issues/267